From ffe7cf672fb49e9cfba96b568a528b67dc1f8913 Mon Sep 17 00:00:00 2001
From: Sagi Maimon <smaimon@habana.ai>
Date: Tue, 02 Feb 2021 11:43:23 +0200
Subject: [PATCH] [SW-29734] preboot: implementing strncpy function

corrupted fit information can cause stack based buffer overflow.
adding strncpy implementation for testing string size.

Change-Id: Ifb11b3392476433da0eab913aae2cd6a4129b094
Signed-off-by: Sagi Maimon <smaimon@habana.ai>
---

diff --git a/common/string.c b/common/string.c
index 4b3b55d..08dce41 100644
--- a/common/string.c
+++ b/common/string.c
@@ -201,3 +201,22 @@
        return NULL;
 }
 
+/**
+ * strncpy - Copy a length-limited, %NUL-terminated string
+ * @dest: Where to copy the string to
+ * @src: Where to copy the string from
+ * @count: The maximum number of bytes to copy
+ *
+ * Note that unlike userspace strncpy, this does not %NUL-pad the buffer.
+ * However, the result is not %NUL-terminated if the source exceeds
+ * @count bytes.
+ */
+char * strncpy(char * dest,const char *src, size_t count)
+{
+	char *tmp = dest;
+
+	while (count-- && (*dest++ = *src++) != '\0')
+		/* nothing */;
+
+	return tmp;
+}
diff --git a/include/string.h b/include/string.h
index 6c97643..f4e61d5 100644
--- a/include/string.h
+++ b/include/string.h
@@ -20,6 +20,7 @@
 
 int memcmp(const void * cs,const void * ct, size_t count);
 int strcmp(const char * cs,const char * ct);
+char *strncpy(char *dest, const char *src, size_t count);
 
 size_t strlen(const char * s);
 size_t strnlen(const char * s, size_t count);
